core/templates/dev/head/exploration_editor/EditorServices.js

```diff
@@ -1952,3 +2012,302 @@ oppia.factory('explorationWarningsService', [
     };
   }
 ]);
+
+oppia.factory('lostChangesService', ['utilsService', function(utilsService) {
+  var CMD_ADD_STATE = 'add_state';
+  var CMD_RENAME_STATE = 'rename_state';
+  var CMD_DELETE_STATE = 'delete_state';
+  var CMD_EDIT_STATE_PROPERTY = 'edit_state_property';
+
+  var makeRulesListHumanReadable = function(answerGroupValue) {
+    var rulesList = [];
+    answerGroupValue.rule_specs.forEach(function(ruleSpec) {
+      var ruleElm = angular.element('<div></div>');
+      ruleElm.append('<p>Type: ' + ruleSpec.rule_type + '</p>');
+      ruleElm.append(
+        '<p>Value: ' +
+          (
+            Object.keys(ruleSpec.inputs).map(function(input) {
+              return ruleSpec.inputs[input];
+            })
+          ).toString() +
+          '</p>');
+      rulesList.push(ruleElm);
+    });
+
+    return rulesList;
+  };
+
+  // An edit is represented either as an object or an array. If it's an object,
+  // then simply return that object. In case of an array, return the last item
+  var getStatePropertyValue = function(statePropertyValue) {
+    return angular.isArray(statePropertyValue) ?
+      statePropertyValue[statePropertyValue.length - 1] : statePropertyValue;
+  };
+
+  // Detects whether an object of the type 'answer_group' or 'default_outcome'
+  // has been added, edited or deleted.
+  // Returns - 'addded', 'edited' or 'deleted' accordingly.
+  var getRelativeChangeToGroups = function(changeObject) {
+    var newValue = changeObject.new_value;
+    var oldValue = changeObject.old_value;
+    var result = '';
+
+    if (angular.isArray(newValue) && angular.isArray(oldValue)) {
+      result = (newValue.length > oldValue.length) ?
+        'added' : (newValue.length === oldValue.length) ?
+        'edited' : 'deleted';
+    } else {
+      if (!utilsService.isEmpty(oldValue)) {
+        if (!utilsService.isEmpty(newValue)) {
+          result = 'edited';
+        } else {
+          result = 'deleted';
+        }
+      } else if (!utilsService.isEmpty(newValue)) {
+        result = 'added';
+      }
+    }
+    return result;
+  };
+
+  var makeHumanReadable = function(lostChanges) {
+    var outerHtml = angular.element('<ul></ul>');
+    var stateWiseEditsMapping = {};
+    // The variable stateWiseEditsMapping stores the edits grouped by state.
+    // For instance, you made the following edits:
+    // 1. Changed content to 'Welcome' instead of '' in 'First Card'.
+    // 2. Added an interaction in this state.
+    // 2. Added a new state 'End'.
+    // 3. Ended Exporation from state 'End'.
+    // stateWiseEditsMapping will look something like this:
+    // - 'First Card': [
+    //   - 'Edited Content: Welcome!',:
+    //   - 'Added Interaction: Continue',
+    //   - 'Added interaction customizations']
+    // - 'End': ['Ended exploration']
+
+    lostChanges.forEach(function(lostChange) {
+      var lostChangeEl = angular.element('<li></li>');
```

Comment:
I think you're now just using lostChangeEl and lostChangeValue locally within cases (which is good) so there's no need to create them here as that will just have the reader wondering what they are for.